server/src/main/java/org/opensearch/search/SearchService.java

```diff
@@ -793,28 +863,57 @@ final ReaderContext createAndPutReaderContext(
         shard.awaitShardSearchActive(ignored -> {
+            Releasable decreasePitContexts = null;
             Engine.SearcherSupplier searcherSupplier = null;
             ReaderContext readerContext = null;
+            boolean success = false;
             try {
+                // use this when reader context is freed
+                decreasePitContexts = openPitContexts::decrementAndGet;
+                if (openPitContexts.incrementAndGet() > maxOpenPitContext) {
+                    throw new OpenSearchRejectedExecutionException(
+                        "Trying to create too many Point In Time contexts. Must be less than or equal to: ["
+                            + maxOpenPitContext
+                            + "]. "
+                            + "This limit can be set by changing the ["
+                            + MAX_OPEN_PIT_CONTEXT.getKey()
+                            + "] setting."
+                    );
+                }
                 searcherSupplier = shard.acquireSearcherSupplier();
                 final ShardSearchContextId id = new ShardSearchContextId(sessionId, idGenerator.incrementAndGet());
-                readerContext = new ReaderContext(id, indexService, shard, searcherSupplier, keepAlive.millis(), false);
+                readerContext = new PitReaderContext(id, indexService, shard, searcherSupplier, keepAlive.millis(), false);
                 final ReaderContext finalReaderContext = readerContext;
                 searcherSupplier = null; // transfer ownership to reader context
+
                 searchOperationListener.onNewReaderContext(readerContext);
-                readerContext.addOnClose(() -> searchOperationListener.onFreeReaderContext(finalReaderContext));
+                searchOperationListener.onNewPitContext(finalReaderContext);
+                readerContext.addOnClose(decreasePitContexts);
+                decreasePitContexts = null;
+
+                readerContext.addOnClose(() -> {
+                    searchOperationListener.onFreeReaderContext(finalReaderContext);
+                    searchOperationListener.onFreePitContext(finalReaderContext);
+                });
+                // add the newly created pit reader context to active readers
                 putReaderContext(readerContext);
                 readerContext = null;
                 listener.onResponse(finalReaderContext.id());
+                success = true;
             } catch (Exception exc) {
-                Releasables.closeWhileHandlingException(searcherSupplier, readerContext);
                 listener.onFailure(exc);
+            } finally {
+                if (success) {
+                    Releasables.close(readerContext, searcherSupplier, decreasePitContexts);
+                } else {
+                    Releasables.closeWhileHandlingException(searcherSupplier, readerContext, decreasePitContexts);
+                }
```

Comment:
Lets avoid an explicit close on null reference and move `closeWhileHandlingException ` within the exception block